File: sdk/python/pact/delegation.py
"""Signed delegation chains for the Pact protocol.

A delegation is a signed statement from one entity (delegator) granting
specific capabilities to another entity (delegate). Chains of delegations
form a trust path from a root authority (human) to the acting agent.
"""

from __future__ import annotations

import secrets
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Optional

from .canonical import canonical_json
from .capabilities import CapabilitySet, parse_capability_set
from .encoding import decode_base64url, encode_base64url
from .identity import Identity


@dataclass
class DelegationConstraints:
    """Bounds on a delegation."""

    expires: str
    """RFC 3339 timestamp after which the delegation is invalid."""

    max_chain_depth: int
    """Maximum total chain length. 1 = no sub-delegation."""

    not_before: str = ""
    """RFC 3339 timestamp before which the delegation is invalid (optional)."""

    def to_dict(self) -> dict:
        d: dict = {
            "expires": self.expires,
            "max_chain_depth": self.max_chain_depth,
        }
        if self.not_before:
            d["not_before"] = self.not_before
        return d


@dataclass
class DelegationEntity:
    """Public identity info embedded in a delegation."""

    type: str
    id: str
    public_key: str
    name: str = ""

    def to_dict(self) -> dict:
        d: dict = {
            "type": self.type,
            "id": self.id,
            "public_key": self.public_key,
        }
        if self.name:
            d["name"] = self.name
        return d

    @classmethod
    def from_dict(cls, d: dict) -> DelegationEntity:
        return cls(
            type=d["type"],
            id=d["id"],
            public_key=d["public_key"],
            name=d.get("name", ""),
        )


@dataclass
class Delegation:
    """A signed authorization from one entity to another.

    The core primitive of Pact — the cryptographic pact between delegator and delegate.
    """

    type: str = "delegation"
    version: str = "1.0"
    id: str = ""
    from_entity: DelegationEntity = field(
        default_factory=lambda: DelegationEntity("", "", ""))
    to_entity: DelegationEntity = field(
        default_factory=lambda: DelegationEntity("", "", ""))
    capabilities: list[str] = field(default_factory=list)
    constraints: DelegationConstraints = field(
        default_factory=lambda: DelegationConstraints("", 1)
    )
    issued_at: str = ""
    signature: str = ""

    def to_dict(self) -> dict:
        """Serialize to dict for JSON/canonical JSON."""
        d: dict = {
            "type": self.type,
            "version": self.version,
            "id": self.id,
            "from": self.from_entity.to_dict(),
            "to": self.to_entity.to_dict(),
            "capabilities": self.capabilities,
            "constraints": self.constraints.to_dict(),
            "issued_at": self.issued_at,
            "signature": self.signature,
        }
        return d

    @classmethod
    def from_dict(cls, d: dict) -> Delegation:
        """Deserialize from dict."""
        constraints_d = d.get("constraints", {})
        return cls(
            type=d.get("type", "delegation"),
            version=d.get("version", "1.0"),
            id=d.get("id", ""),
            from_entity=DelegationEntity.from_dict(d["from"]),
            to_entity=DelegationEntity.from_dict(d["to"]),
            capabilities=d.get("capabilities", []),
            constraints=DelegationConstraints(
                expires=constraints_d.get("expires", ""),
                max_chain_depth=constraints_d.get("max_chain_depth", 1),
                not_before=constraints_d.get("not_before", ""),
            ),
            issued_at=d.get("issued_at", ""),
            signature=d.get("signature", ""),
        )

    def signing_payload(self) -> bytes:
        """Return the canonical JSON bytes that are signed.

        Sets signature to "" before canonicalization (matching Go implementation).
        """
        d = self.to_dict()
        d["signature"] = ""
        return canonical_json(d)

    def _sign(self, signer: Identity) -> None:
        """Compute and set the Ed25519 signature."""
        self.signature = ""
        payload = self.signing_payload()
        sig = signer.sign(payload)
        self.signature = encode_base64url(sig)


# Type alias for clarity
DelegationChain = list[Delegation]


def new_delegation(
    from_identity: Identity,
    to_identity: Identity,
    capabilities: list[str],
    ttl: timedelta,
    max_chain_depth: int = 1,
    not_before: Optional[datetime] = None,
    delegation_id: Optional[str] = None,
) -> Delegation:
    """Create and sign a new delegation.

    The delegator (from_identity) signs the delegation with their private key,
    cryptographically binding themselves to the authorization.

    Args:
        from_identity: The delegator (must have a private key).
        to_identity: The delegate (only public key needed).
        capabilities: Permission strings being granted.
        ttl: How long the delegation is valid.
        max_chain_depth: Maximum sub-delegation depth. Default 1 (no sub-delegation).
        not_before: Earliest validity time. Default: now.
        delegation_id: Optional explicit delegation ID (for test vectors).

    Returns:
        A signed Delegation.

    Raises:
        ValueError: If required fields are missing or invalid.
    """
    if from_identity is None:
        raise ValueError("Delegation requires a 'from' identity")
    if to_identity is None:
        raise ValueError("Delegation requires a 'to' identity")
    if from_identity.private_key is None:
        raise ValueError("Delegator must have a private key to sign")
    if not capabilities:
        raise ValueError("Delegation requires at least one capability")
    if ttl.total_seconds() <= 0:
        raise ValueError("Delegation requires a positive TTL")

    if max_chain_depth <= 0:
        max_chain_depth = 1

    now = datetime.now(timezone.utc)
    if not_before is None:
        not_before = now

    if delegation_id is None:
        delegation_id = generate_delegation_id()

    d = Delegation(
        type="delegation",
        version="1.0",
        id=delegation_id,
        from_entity=DelegationEntity(
            type=from_identity.type,
            id=from_identity.id,
            public_key=from_identity.public_key,
            name=from_identity.name,
        ),
        to_entity=DelegationEntity(
            type=to_identity.type,
            id=to_identity.id,
            public_key=to_identity.public_key,
            name=to_identity.name,
        ),
        capabilities=capabilities,
        constraints=DelegationConstraints(
            expires=(now + ttl).strftime("%Y-%m-%dT%H:%M:%SZ"),
            max_chain_depth=max_chain_depth,
            not_before=not_before.strftime("%Y-%m-%dT%H:%M:%SZ"),
        ),
        issued_at=now.strftime("%Y-%m-%dT%H:%M:%SZ"),
        signature="",
    )

    d._sign(from_identity)
    return d


def sub_delegate(
    chain: DelegationChain,
    agent: Identity,
    to: Identity,
    capabilities: list[str],
    ttl: timedelta,
) -> tuple[Delegation, DelegationChain]:
    """Create a sub-delegation from the chain's terminal agent to a new agent.

    The new delegation's capabilities must be a subset of the parent's.
    The chain depth must not exceed the root's max_chain_depth.

    Args:
        chain: The existing delegation chain.
        agent: The terminal agent in the chain (must have private key).
        to: The new delegate.
        capabilities: Capabilities to grant (must be subset of parent's).
        ttl: TTL for the sub-delegation.

    Returns:
        Tuple of (new delegation, extended chain).

    Raises:
        ValueError: If constraints are violated.
    """
    if not chain:
        raise ValueError("Cannot sub-delegate from empty chain")

    last = chain[-1]

    # Verify the agent is the terminal entity
    if agent.id != last.to_entity.id:
        raise ValueError(
            f"Agent identity {agent.id} does not match chain terminal {last.to_entity.id}"
        )

    # Check chain depth
    root = chain[0]
    if len(chain) >= root.constraints.max_chain_depth:
        raise ValueError(
            f"Sub-delegation would exceed max chain depth {root.constraints.max_chain_depth}"
        )

    # Verify narrowing
    parent_caps = parse_capability_set(last.capabilities)
    child_caps = parse_capability_set(capabilities)
    if not parent_caps.covers(child_caps):
        raise ValueError(
            "Sub-delegation capabilities exceed parent scope (narrowing-only violation)"
        )

    # TTL cannot exceed parent's remaining TTL
    parent_expires = datetime.fromisoformat(last.constraints.expires)
    if parent_expires.tzinfo is None:
        parent_expires = parent_expires.replace(tzinfo=timezone.utc)
    remaining = parent_expires - datetime.now(timezone.utc)
    if ttl > remaining:
        ttl = remaining

    # Clamp max chain depth
    new_max_depth = root.constraints.max_chain_depth - len(chain)
    if new_max_depth < 1:
        new_max_depth = 1

    d = new_delegation(
        from_identity=agent,
        to_identity=to,
        capabilities=capabilities,
        ttl=ttl,
        max_chain_depth=new_max_depth,
    )

    new_chain = chain + [d]
    return d, new_chain


def chain_terminal_entity(chain: DelegationChain) -> str:
    """Return the identity ID of the last delegate in the chain."""
    if not chain:
        return ""
    return chain[-1].to_entity.id


def chain_root_authority(chain: DelegationChain) -> str:
    """Return the identity ID of the root delegator."""
    if not chain:
        return ""
    return chain[0].from_entity.id


def generate_delegation_id() -> str:
    """Generate a random delegation ID: 'd-' + 32 hex chars."""
    return "d-" + secrets.token_hex(16)

File: delegation.go
package pact

import (
	"crypto/rand"
	"fmt"
	"time"
)

// Delegation represents a signed authorization from one entity to another.
// It is the core primitive of Pact — the cryptographic pact between delegator and delegate.
//
// A delegation says: "I (from) authorize you (to) to do these things (capabilities)
// within these bounds (constraints), and I sign this with my private key."
type Delegation struct {
	// Type is always "delegation".
	Type string `json:"type"`

	// Version is the protocol version.
	Version string `json:"version"`

	// ID is a unique identifier for this delegation.
	ID string `json:"id"`

	// From is the delegator's public identity.
	From DelegationEntity `json:"from"`

	// To is the delegate's public identity.
	To DelegationEntity `json:"to"`

	// Capabilities are the permissions being granted.
	Capabilities []string `json:"capabilities"`

	// Constraints bound the delegation.
	Constraints DelegationConstraints `json:"constraints"`

	// IssuedAt is when the delegation was created.
	IssuedAt string `json:"issued_at"`

	// Signature is the delegator's Ed25519 signature over the canonical form
	// of this delegation (with signature field empty).
	Signature string `json:"signature"`
}

// DelegationEntity is the public identity info embedded in a delegation.
type DelegationEntity struct {
	Type      EntityType `json:"type"`
	ID        string     `json:"id"`
	PublicKey string     `json:"public_key"`
	Name      string     `json:"name,omitempty"`
}

// DelegationConstraints define the bounds of a delegation.
type DelegationConstraints struct {
	// Expires is when this delegation becomes invalid (RFC 3339).
	Expires string `json:"expires"`

	// MaxChainDepth limits how many times this delegation can be sub-delegated.
	// 1 = no sub-delegation allowed. 2 = one level of sub-delegation. etc.
	MaxChainDepth int `json:"max_chain_depth"`

	// NotBefore is the earliest time this delegation is valid (optional, RFC 3339).
	NotBefore string `json:"not_before,omitempty"`
}

// DelegationChain is an ordered list of delegations forming a trust path
// from a root authority (typically human) to the acting agent.
//
// Chain order: [root delegation, sub-delegation, sub-sub-delegation, ...]
// The first delegation's "from" is the root authority.
// The last delegation's "to" is the acting agent.
type DelegationChain []Delegation

// DelegationRequest specifies the parameters for creating a new delegation.
type DelegationRequest struct {
	// From is the delegator (must have a private key for signing).
	From *Identity

	// To is the delegate (only public key needed).
	To *Identity

	// Capabilities are the permissions being granted.
	Capabilities []string

	// TTL is how long the delegation is valid.
	TTL time.Duration

	// MaxChainDepth limits sub-delegation depth. 0 means use default (1 = no sub-delegation).
	MaxChainDepth int

	// NotBefore is the earliest validity time. Zero means now.
	NotBefore time.Time
}

// NewDelegation creates and signs a new delegation.
//
// The delegator (from) signs the delegation with their private key,
// cryptographically binding themselves to the authorization.
func NewDelegation(req DelegationRequest) (*Delegation, error) {
	if req.From == nil {
		return nil, fmt.Errorf("pact: delegation requires a 'from' identity")
	}
	if req.To == nil {
		return nil, fmt.Errorf("pact: delegation requires a 'to' identity")
	}
	if req.From.PrivateKey == nil {
		return nil, fmt.Errorf("pact: delegator must have a private key to sign")
	}
	if len(req.Capabilities) == 0 {
		return nil, fmt.Errorf("pact: delegation requires at least one capability")
	}
	if req.TTL <= 0 {
		return nil, fmt.Errorf("pact: delegation requires a positive TTL")
	}

	maxDepth := req.MaxChainDepth
	if maxDepth <= 0 {
		maxDepth = 1
	}

	now := time.Now().UTC()
	notBefore := req.NotBefore
	if notBefore.IsZero() {
		notBefore = now
	}

	id, err := generateDelegationID()
	if err != nil {
		return nil, err
	}

	d := &Delegation{
		Type:    "delegation",
		Version: "1.0",
		ID:      id,
		From: DelegationEntity{
			Type:      req.From.Type,
			ID:        req.From.ID,
			PublicKey: req.From.PublicKey,
			Name:      req.From.Name,
		},
		To: DelegationEntity{
			Type:      req.To.Type,
			ID:        req.To.ID,
			PublicKey: req.To.PublicKey,
			Name:      req.To.Name,
		},
		Capabilities: req.Capabilities,
		Constraints: DelegationConstraints{
			Expires:       now.Add(req.TTL).Format(time.RFC3339),
			MaxChainDepth: maxDepth,
			NotBefore:     notBefore.Format(time.RFC3339),
		},
		IssuedAt:  now.Format(time.RFC3339),
		Signature: "",
	}

	// Sign the delegation
	if err := d.sign(req.From); err != nil {
		return nil, err
	}

	return d, nil
}

// SubDelegate creates a new delegation from this chain's terminal agent to a new agent.
// The new delegation's capabilities must be a subset of the parent's.
// The chain depth must not exceed the root's max_chain_depth.
func (chain DelegationChain) SubDelegate(agent *Identity, to *Identity, capabilities []string, ttl time.Duration) (*Delegation, DelegationChain, error) {
	if len(chain) == 0 {
		return nil, nil, fmt.Errorf("pact: cannot sub-delegate from empty chain")
	}

	last := chain[len(chain)-1]

	// Verify the agent is the terminal entity in the chain
	if agent.ID != last.To.ID {
		return nil, nil, fmt.Errorf("pact: agent identity %s does not match chain terminal %s", agent.ID, last.To.ID)
	}

	// Check chain depth
	root := chain[0]
	if len(chain) >= root.Constraints.MaxChainDepth {
		return nil, nil, fmt.Errorf("pact: sub-delegation would exceed max chain depth %d", root.Constraints.MaxChainDepth)
	}

	// Verify narrowing: new capabilities must be covered by parent capabilities
	parentCaps, err := ParseCapabilitySet(last.Capabilities)
	if err != nil {
		return nil, nil, fmt.Errorf("pact: invalid parent capabilities: %w", err)
	}
	childCaps, err := ParseCapabilitySet(capabilities)
	if err != nil {
		return nil, nil, fmt.Errorf("pact: invalid child capabilities: %w", err)
	}
	if !parentCaps.Covers(childCaps) {
		return nil, nil, fmt.Errorf("pact: sub-delegation capabilities exceed parent scope (narrowing-only violation)")
	}

	// TTL cannot exceed parent's remaining TTL
	parentExpires, err := time.Parse(time.RFC3339, last.Constraints.Expires)
	if err != nil {
		return nil, nil, fmt.Errorf("pact: invalid parent expiry: %w", err)
	}
	remaining := time.Until(parentExpires)
	if ttl > remaining {
		ttl = remaining
	}

	// Max chain depth for sub-delegation is parent's minus current depth
	newMaxDepth := root.Constraints.MaxChainDepth - len(chain)
	if newMaxDepth < 1 {
		newMaxDepth = 1
	}

	// Create the sub-delegation
	d, err := NewDelegation(DelegationRequest{
		From:          agent,
		To:            to,
		Capabilities:  capabilities,
		TTL:           ttl,
		MaxChainDepth: newMaxDepth,
	})
	if err != nil {
		return nil, nil, fmt.Errorf("pact: sub-delegation failed: %w", err)
	}

	// Build the extended chain
	newChain := make(DelegationChain, len(chain)+1)
	copy(newChain, chain)
	newChain[len(chain)] = *d

	return d, newChain, nil
}

// TerminalEntity returns the identity ID of the last delegate in the chain
// (the entity currently authorized to act).
func (chain DelegationChain) TerminalEntity() string {
	if len(chain) == 0 {
		return ""
	}
	return chain[len(chain)-1].To.ID
}

// RootAuthority returns the identity ID of the root delegator
// (typically the human who started the chain).
func (chain DelegationChain) RootAuthority() string {
	if len(chain) == 0 {
		return ""
	}
	return chain[0].From.ID
}

// sign computes and sets the Ed25519 signature over the canonical form of the delegation.
func (d *Delegation) sign(signer *Identity) error {
	// Temporarily clear signature for signing
	d.Signature = ""

	payload, err := canonicalJSON(d)
	if err != nil {
		return fmt.Errorf("pact: failed to canonicalize delegation: %w", err)
	}

	sig, err := signer.Sign(payload)
	if err != nil {
		return err
	}

	d.Signature = encodeBase64URL(sig)
	return nil
}

// signingPayload returns the canonical bytes that were signed.
func (d *Delegation) signingPayload() ([]byte, error) {
	dCopy := *d
	dCopy.Signature = ""
	return canonicalJSON(&dCopy)
}

// generateDelegationID creates a random delegation ID.
func generateDelegationID() (string, error) {
	b := make([]byte, 16)
	_, err := randRead(b)
	if err != nil {
		return "", fmt.Errorf("pact: failed to generate delegation ID: %w", err)
	}
	return fmt.Sprintf("d-%x", b), nil
}

// randRead is a variable for testing — reads random bytes.
var randRead = randReadDefault

func randReadDefault(b []byte) (int, error) {
	return randReaderRead(b)
}

// We use a separate function to avoid import cycle issues with testing
func randReaderRead(b []byte) (int, error) {
	return _randReader.Read(b)
}

// _randReader uses crypto/rand
var _randReader = cryptoRandReader{}

type cryptoRandReader struct{}

func (cryptoRandReader) Read(b []byte) (int, error) {
	return randReadFromCryptoRand(b)
}

func randReadFromCryptoRand(b []byte) (int, error) {
	// imported at top: "crypto/rand"
	return rand.Read(b)
}
